src/function/debounce.ts

```diff
@@ -0,0 +1,52 @@
+/**
+ * Return a function which will only call {@link cb} when a delay is passed after its last call
+ * @example
+ * function test(id: number) {
+ *    console.log(id);
+ * }
+ * const debounced = debounce(test, 1000);
+ * debounced(1);
+ * debounced(2);
+ * debounced(3);
+ * debounced(4);
+ * debounced(5);
+ * // will log 5 after 1 second
+ * @param {Function} cb function to be executed after the delay
+ * @param {number} delay delay in milliseconds
+ * @param {boolean} immediate if *true*, will execute {@link cb} immediately
+ * if it hasn't been called in the last {@link delay} milliseconds. Default is *false*
+ *
+ * #### Example
+ * ```ts
+ * function test(id: number) {
+ *    console.log(id);
+ * }
+ * const debounced = debounce(test, 1000, true);
+ * debounced(1); // immediately logs 1
+ * debounced(2);
+ * debounced(3);
+ * debounced(4);
+ * debounced(5); // will log 5 after 1 second
+ * ```
+ *
+ * @returns {Function} a wrapper function
+ *
+ */
+export function debounce<F extends (...args: any[]) => any>(
```

Comment:
I'd say  `debounce({ delay: number; isImmediate?: boolean }, f)` is better as we can curry like:

```ts
const d1Sec = debounce({ delay : 1000, isImidiate: false })
d1Sec(f)
d1Sec(g)
```

Reply:
I think I'm missing something here. How is `d1Sec` callable with a function?
Also if the signature is `debounce({ delay: number; isImmediate?: boolean }, f)` how could we call it like `debounce({ delay : 1000, isImidiate: false })` ?

Reply:
That was a pseudo code.
The actual code is:

```
const curriedDebounce = curry(debounce)
const d1Sec = curriedDebounce({ delay : 1000, isImidiate: false })
d1Sec(f)
d1Sec(g)
```

Reply:
My bad, I did not know about curries